File: frontend/lib/permissions.ts
/**
 * Client-side mirror of backend ROLE_PERMISSIONS.
 * Used for UI-level gating (hiding buttons, menu items, etc.)
 * The backend is the source of truth — this is for UX only.
 */

export type UserRole = "admin" | "project_manager" | "translator" | "reviewer" | "curator" | "viewer";

export type Permission =
  | "document:read" | "document:create" | "document:update" | "document:delete"
  | "translate:execute" | "translate:upload"
  | "segment:read" | "segment:edit"
  | "review:approve" | "review:reject"
  | "knowledge:read" | "knowledge:manage"
  | "audit:read" | "audit:export"
  | "tools:use"
  | "users:read" | "users:manage" | "system:admin";

const ALL_PERMISSIONS: Permission[] = [
  "document:read", "document:create", "document:update", "document:delete",
  "translate:execute", "translate:upload",
  "segment:read", "segment:edit",
  "review:approve", "review:reject",
  "knowledge:read", "knowledge:manage",
  "audit:read", "audit:export",
  "tools:use",
  "users:read", "users:manage", "system:admin",
];

export const ROLE_PERMISSIONS: Record<UserRole, Set<Permission>> = {
  admin: new Set(ALL_PERMISSIONS),

  project_manager: new Set([
    "document:read", "document:create", "document:update", "document:delete",
    "translate:execute", "translate:upload",
    "segment:read", "segment:edit",
    "review:approve", "review:reject",
    "knowledge:read", "knowledge:manage",
    "audit:read", "audit:export",
    "tools:use", "users:read",
  ]),

  translator: new Set([
    "document:read", "document:create", "document:update",
    "translate:execute", "translate:upload",
    "segment:read", "segment:edit",
    "knowledge:read",
    "audit:read",
    "tools:use",
  ]),

  reviewer: new Set([
    "document:read",
    "segment:read",
    "review:approve", "review:reject",
    "knowledge:read",
    "audit:read", "audit:export",
    "tools:use",
  ]),

  curator: new Set([
    "document:read",
    "segment:read",
    "knowledge:read", "knowledge:manage",
    "audit:read",
    "tools:use",
  ]),

  viewer: new Set([
    "document:read",
    "segment:read",
    "knowledge:read",
    "audit:read",
    "tools:use",
  ]),
};

export function hasPermission(role: UserRole, permission: Permission): boolean {
  return ROLE_PERMISSIONS[role]?.has(permission) ?? false;
}

export function hasRole(userRole: UserRole, allowedRoles: UserRole[]): boolean {
  return allowedRoles.includes(userRole);
}

export const ROLE_LABELS: Record<UserRole, string> = {
  admin: "Admin",
  project_manager: "Project Manager",
  translator: "Translator",
  reviewer: "Reviewer",
  curator: "Curator",
  viewer: "Viewer",
};

export const ROLE_COLORS: Record<UserRole, string> = {
  admin: "#ea4335",
  project_manager: "#4285f4",
  translator: "#34a853",
  reviewer: "#fbbc04",
  curator: "#9334e6",
  viewer: "#5f6368",
};

File: app/auth/permissions.py
"""
Permission & Role definitions for TransMax RBAC.

Flat permission matrix with 6 pharma-appropriate roles.
"""
from enum import Enum
from typing import Set, Dict


class Permission(str, Enum):
    # Documents
    DOCUMENT_READ = "document:read"
    DOCUMENT_CREATE = "document:create"
    DOCUMENT_UPDATE = "document:update"
    DOCUMENT_DELETE = "document:delete"

    # Translation
    TRANSLATE_EXECUTE = "translate:execute"
    TRANSLATE_UPLOAD = "translate:upload"

    # Segments
    SEGMENT_READ = "segment:read"
    SEGMENT_EDIT = "segment:edit"

    # Review & Approval
    REVIEW_APPROVE = "review:approve"
    REVIEW_REJECT = "review:reject"

    # Knowledge / Black Book
    KNOWLEDGE_READ = "knowledge:read"
    KNOWLEDGE_MANAGE = "knowledge:manage"
    # TMX-3045 / Pillar 1: signed promotion of a learned rule from PROPOSED
    # to ACTIVE. Strictly stronger than KNOWLEDGE_MANAGE — KNOWLEDGE_MANAGE
    # lets a curator create or edit rules manually, but ONLY a holder of
    # RULE_APPROVE may flip a learned (LLM-extracted) rule to ACTIVE.
    RULE_APPROVE = "rule:approve"

    # Audit
    AUDIT_READ = "audit:read"
    AUDIT_EXPORT = "audit:export"

    # Tools
    TOOLS_USE = "tools:use"

    # Admin
    USERS_READ = "users:read"
    USERS_MANAGE = "users:manage"
    SYSTEM_ADMIN = "system:admin"


class UserRole(str, Enum):
    ADMIN = "admin"
    PROJECT_MANAGER = "project_manager"
    TRANSLATOR = "translator"
    REVIEWER = "reviewer"
    CURATOR = "curator"
    VIEWER = "viewer"


# Flat permission matrix
ROLE_PERMISSIONS: Dict[UserRole, Set[Permission]] = {
    UserRole.ADMIN: set(Permission),  # All permissions

    UserRole.PROJECT_MANAGER: {
        Permission.DOCUMENT_READ, Permission.DOCUMENT_CREATE,
        Permission.DOCUMENT_UPDATE, Permission.DOCUMENT_DELETE,
        Permission.TRANSLATE_EXECUTE, Permission.TRANSLATE_UPLOAD,
        Permission.SEGMENT_READ, Permission.SEGMENT_EDIT,
        Permission.REVIEW_APPROVE, Permission.REVIEW_REJECT,
        Permission.KNOWLEDGE_READ, Permission.KNOWLEDGE_MANAGE,
        Permission.RULE_APPROVE,  # TMX-3045: PMs may sign rule promotions.
        Permission.AUDIT_READ, Permission.AUDIT_EXPORT,
        Permission.TOOLS_USE, Permission.USERS_READ,
    },

    UserRole.TRANSLATOR: {
        Permission.DOCUMENT_READ, Permission.DOCUMENT_CREATE,
        Permission.DOCUMENT_UPDATE,
        Permission.TRANSLATE_EXECUTE, Permission.TRANSLATE_UPLOAD,
        Permission.SEGMENT_READ, Permission.SEGMENT_EDIT,
        Permission.KNOWLEDGE_READ,
        Permission.AUDIT_READ,
        Permission.TOOLS_USE,
    },

    UserRole.REVIEWER: {
        Permission.DOCUMENT_READ,
        Permission.SEGMENT_READ,
        Permission.REVIEW_APPROVE, Permission.REVIEW_REJECT,
        Permission.KNOWLEDGE_READ,
        Permission.AUDIT_READ, Permission.AUDIT_EXPORT,
        Permission.TOOLS_USE,
    },

    UserRole.CURATOR: {
        Permission.DOCUMENT_READ,
        Permission.SEGMENT_READ,
        Permission.KNOWLEDGE_READ, Permission.KNOWLEDGE_MANAGE,
        Permission.RULE_APPROVE,  # TMX-3045: curators are the canonical rule approvers.
        Permission.AUDIT_READ,
        Permission.TOOLS_USE,
    },

    UserRole.VIEWER: {
        Permission.DOCUMENT_READ,
        Permission.SEGMENT_READ,
        Permission.KNOWLEDGE_READ,
        Permission.AUDIT_READ,
        Permission.TOOLS_USE,
    },
}


def has_permission(role: UserRole, permission: Permission) -> bool:
    """Check if a role has a specific permission."""
    return permission in ROLE_PERMISSIONS.get(role, set())
